db/commitlog/commitlog.cc

```diff
@@ -1511,7 +1511,9 @@ future<db::commitlog::segment_manager::sseg_ptr> db::commitlog::segment_manager:
     if (!cfg.allow_going_over_size_limit && max_disk_size != 0 && totals.total_size_on_disk >= max_disk_size) {
         clogger.debug("Disk usage ({} MB) exceeds maximum ({} MB) - allocation will wait...", totals.total_size_on_disk/(1024*1024), max_disk_size/(1024*1024));
         auto f = cfg.reuse_segments ? _recycled_segments.not_empty() :  _disk_deletions.get_shared_future();
-        return f.then([this] {
+        return f.handle_exception([this](auto ep) {
+            clogger.warn("Exception while waiting for segments {}. Will retry allocation...", ep);
+        }).then([this] {
             return allocate_segment();
         });
```

Comment:
If the same error happens, won't we spin in this function?

Reply:
Assuming a segment eventually is made available, we will resolve at that time. If disk errors just keep happening, we will indeed deadlock again, just like before.
I guess we could add a retry-counter or something. Not sure about adding it now though. Technically, we should try to adhere to any timeouts specified on higher levels, but our waiting with timeout tends to abandon the timed out future, which is not great either...